src/clr-core/internal/base/focus-trap.base.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { html, LitElement } from 'lit-element';
+import { property } from '../decorators/property';
+import { FocusTrapTracker } from '../services/focus-trap-tracker.service';
+import { addAttributeValue, isHTMLElement, removeAttributeValue } from '../utils/dom';
+import { isBrowser } from '../utils/exists';
+
+export class CdsBaseFocusTrap extends LitElement {
+  private previousFocus: Element;
+  private topReboundEl: any;
+  private bottomReboundEl: any;
+  private onFocusInEvent: any;
+
+  @property({ attribute: false })
+  private hostElement = this;
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    if (document.activeElement) {
+      this.previousFocus = document.activeElement;
+    }
+
+    // add attribute to prevent scroll on body
+    addAttributeValue(document.body, 'cds-layout', 'no-scrolling');
+
+    this.addReboundElements();
+
+    FocusTrapTracker.setCurrent(this.hostElement);
+    this.hostElement.setAttribute('tabindex', '0');
+    this.hostElement.focus();
+
+    this.onFocusInEvent = this.onFocusIn.bind(this);
+    document.addEventListener('focusin', this.onFocusInEvent);
+  }
+
+  disconnectedCallback() {
+    document.removeEventListener('focusin', this.onFocusInEvent);
+
+    // remove attribute that prevent scroll on body
+    removeAttributeValue(document.body, 'cds-layout', 'no-scrolling');
+
+    this.removeReboundElements();
+    FocusTrapTracker.activatePreviousCurrent();
+
+    if (isHTMLElement(this.previousFocus)) {
+      (this.previousFocus as HTMLElement).focus();
+    }
+
+    super.disconnectedCallback();
+  }
+
+  protected onFocusIn($event: any) {
+    if (FocusTrapTracker.getCurrent() !== this.hostElement) {
+      return;
+    }
+
+    if ($event.target && this.hostElement) {
+      if (
+        !this.hostElement.contains($event.target) ||
+        $event.target === this.topReboundEl ||
+        $event.target === this.bottomReboundEl
```

Comment:
We don't need these checks right? Because the rebounder elements are outside of the host element so the first check will cover these cases too. In our Angular version, we have these check as we have strict focus trap config.

Reply:
Rebound elements are inside the host element in modal actually so it won't pass with the first check.

Reply:
We cannot have the rebound elements inside the host element though. If they are outside, they will longer be able to catch the outside focus and bounce it back into the host element.